File: web/src/lib/terminal/protocol.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2026 Tencent. All rights reserved.

export const TERMINAL_SUBPROTOCOL = 'cube-terminal.v1';
export const TERMINAL_GRANT_PREFIX = 'cube-grant.';

export const TERMINAL_CHANNEL = {
  stdin: 0x00,
  stdout: 0x01,
  stderr: 0x02,
  status: 0x03,
  resize: 0x04,
} as const;

export const MAX_TERMINAL_PAYLOAD_BYTES = 64 * 1024;
export const MAX_TERMINAL_STATUS_BYTES = 4 * 1024;

export const TERMINAL_ERROR_CODES = [
  'TARGET_NOT_FOUND',
  'TARGET_NOT_RUNNING',
  'LIMIT_EXCEEDED',
  'PROTOCOL_ERROR',
  'SHELL_NOT_FOUND',
  'SLOW_PRODUCER',
  'SLOW_CONSUMER',
  'SESSION_LOST',
  'INTERNAL',
  'SERVER_DRAINING',
  'SANDBOX_TRANSITION',
] as const;

export const TERMINAL_CLOSE_REASONS = [
  'USER_CLOSED',
  'IDLE_TIMEOUT',
  'MAX_LIFETIME',
  'SANDBOX_TRANSITION',
  'RUNTIME_EXITED',
  'SESSION_LOST',
  'PROTOCOL_ERROR',
  'SERVER_DRAINING',
  'SLOW_PRODUCER',
  'SLOW_CONSUMER',
  'INTERNAL',
] as const;

export const TERMINAL_REASON_CODES = [
  'UNAUTHORIZED',
  'FORBIDDEN',
  'TARGET_NOT_FOUND',
  'TARGET_NOT_RUNNING',
  'GRANT_INVALID',
  'LIMIT_EXCEEDED',
  'PROTOCOL_ERROR',
  'SHELL_NOT_FOUND',
  'SLOW_PRODUCER',
  'SLOW_CONSUMER',
  'IDLE_TIMEOUT',
  'MAX_LIFETIME',
  'SANDBOX_TRANSITION',
  'RUNTIME_EXITED',
  'SESSION_LOST',
  'INTERNAL',
  'SERVER_DRAINING',
  'USER_CLOSED',
] as const;

export type TerminalErrorCode = (typeof TERMINAL_ERROR_CODES)[number];
export type TerminalCloseReason = (typeof TERMINAL_CLOSE_REASONS)[number];
export type TerminalReasonCode = (typeof TERMINAL_REASON_CODES)[number];

export type TerminalServerFrame =
  | { type: 'stdout'; payload: Uint8Array }
  | {
      type: 'opened';
      sessionId: string;
      replay: { from: number; truncated: boolean };
    }
  | { type: 'exit'; exitCode: number }
  | { type: 'error'; code: TerminalErrorCode }
  | { type: 'close'; reason: TerminalCloseReason };

export type TerminalClientFrame =
  { type: 'stdin'; payload: Uint8Array } | { type: 'resize'; cols: number; rows: number };

export class TerminalProtocolError extends Error {
  constructor(
    readonly closeCode: 1008 | 1009,
    message = 'Invalid terminal protocol frame',
  ) {
    super(message);
    this.name = 'TerminalProtocolError';
  }
}

const errorCodes = new Set<string>(TERMINAL_ERROR_CODES);
const closeReasons = new Set<string>(TERMINAL_CLOSE_REASONS);
const textEncoder = new TextEncoder();

export function assertTerminalDimensions(cols: number, rows: number): void {
  if (
    !Number.isInteger(cols) ||
    !Number.isInteger(rows) ||
    cols < 1 ||
    rows < 1 ||
    cols > 1000 ||
    rows > 1000
  ) {
    throw new TerminalProtocolError(1008, 'Terminal dimensions are out of range');
  }
}

export function encodeResizeFrame(cols: number, rows: number): Uint8Array {
  assertTerminalDimensions(cols, rows);
  const payload = textEncoder.encode(JSON.stringify({ cols, rows }));
  if (payload.byteLength > MAX_TERMINAL_STATUS_BYTES) {
    throw new TerminalProtocolError(1008, 'Terminal resize payload is too large');
  }
  return withChannel(TERMINAL_CHANNEL.resize, payload);
}

export function encodeStdinFrames(data: string | Uint8Array): Uint8Array[] {
  const payload = typeof data === 'string' ? textEncoder.encode(data) : data;
  if (payload.byteLength === 0) {
    return [new Uint8Array([TERMINAL_CHANNEL.stdin])];
  }
  const frames: Uint8Array[] = [];
  for (let offset = 0; offset < payload.byteLength; offset += MAX_TERMINAL_PAYLOAD_BYTES) {
    frames.push(
      withChannel(
        TERMINAL_CHANNEL.stdin,
        payload.subarray(offset, offset + MAX_TERMINAL_PAYLOAD_BYTES),
      ),
    );
  }
  return frames;
}

export function decodeClientFrame(data: unknown): TerminalClientFrame {
  const frame = toBytes(data);
  if (frame.byteLength === 0) {
    throw new TerminalProtocolError(1008, 'Terminal frame is empty');
  }
  const payload = frame.subarray(1);
  if (payload.byteLength > MAX_TERMINAL_PAYLOAD_BYTES) {
    throw new TerminalProtocolError(1009, 'Terminal frame exceeds the payload limit');
  }
  switch (frame[0]) {
    case TERMINAL_CHANNEL.stdin:
      return { type: 'stdin', payload };
    case TERMINAL_CHANNEL.resize: {
      const value = decodeStrictJson(payload, MAX_TERMINAL_STATUS_BYTES);
      assertExactKeys(value, ['cols', 'rows']);
      const cols = value.cols;
      const rows = value.rows;
      if (typeof cols !== 'number' || typeof rows !== 'number') {
        throw new TerminalProtocolError(1008, 'Terminal resize schema is invalid');
      }
      assertTerminalDimensions(cols, rows);
      return { type: 'resize', cols, rows };
    }
    default:
      throw new TerminalProtocolError(1008, 'Terminal client channel is invalid');
  }
}

export function decodeServerFrame(data: unknown): TerminalServerFrame {
  const frame = toBytes(data);
  if (frame.byteLength === 0) {
    throw new TerminalProtocolError(1008, 'Terminal frame is empty');
  }
  const payload = frame.subarray(1);
  if (payload.byteLength > MAX_TERMINAL_PAYLOAD_BYTES) {
    throw new TerminalProtocolError(1009, 'Terminal frame exceeds the payload limit');
  }
  switch (frame[0]) {
    case TERMINAL_CHANNEL.stdout:
      return { type: 'stdout', payload };
    case TERMINAL_CHANNEL.status:
      return decodeStatus(payload);
    default:
      throw new TerminalProtocolError(1008, 'Terminal server channel is invalid');
  }
}

export function isTerminalReasonCode(value: unknown): value is TerminalReasonCode {
  return typeof value === 'string' && (TERMINAL_REASON_CODES as readonly string[]).includes(value);
}

function decodeStatus(payload: Uint8Array): Exclude<TerminalServerFrame, { type: 'stdout' }> {
  const value = decodeStrictJson(payload, MAX_TERMINAL_STATUS_BYTES);
  if (typeof value.type !== 'string') {
    throw new TerminalProtocolError(1008, 'Terminal status type is invalid');
  }
  switch (value.type) {
    case 'opened': {
      assertExactKeys(value, ['type', 'sessionId', 'replay']);
      if (typeof value.sessionId !== 'string' || value.sessionId.length === 0) {
        throw new TerminalProtocolError(1008, 'Terminal opened status is invalid');
      }
      const replay = asRecord(value.replay);
      assertExactKeys(replay, ['from', 'truncated']);
      if (
        typeof replay.from !== 'number' ||
        !Number.isSafeInteger(replay.from) ||
        replay.from < 0 ||
        typeof replay.truncated !== 'boolean'
      ) {
        throw new TerminalProtocolError(1008, 'Terminal replay status is invalid');
      }
      return {
        type: 'opened',
        sessionId: value.sessionId,
        replay: { from: replay.from, truncated: replay.truncated },
      };
    }
    case 'exit': {
      assertExactKeys(value, ['type', 'exitCode']);
      if (
        typeof value.exitCode !== 'number' ||
        !Number.isInteger(value.exitCode) ||
        value.exitCode < -2147483648 ||
        value.exitCode > 2147483647
      ) {
        throw new TerminalProtocolError(1008, 'Terminal exit status is invalid');
      }
      return { type: 'exit', exitCode: value.exitCode };
    }
    case 'error': {
      assertExactKeys(value, ['type', 'code']);
      if (typeof value.code !== 'string' || !errorCodes.has(value.code)) {
        throw new TerminalProtocolError(1008, 'Terminal error code is invalid');
      }
      return { type: 'error', code: value.code as TerminalErrorCode };
    }
    case 'close': {
      assertExactKeys(value, ['type', 'reason']);
      if (typeof value.reason !== 'string' || !closeReasons.has(value.reason)) {
        throw new TerminalProtocolError(1008, 'Terminal close reason is invalid');
      }
      return { type: 'close', reason: value.reason as TerminalCloseReason };
    }
    default:
      throw new TerminalProtocolError(1008, 'Terminal status type is unknown');
  }
}

function decodeStrictJson(payload: Uint8Array, limit: number): Record<string, unknown> {
  if (payload.byteLength === 0 || payload.byteLength > limit) {
    throw new TerminalProtocolError(1008, 'Terminal JSON payload size is invalid');
  }
  let decoded: string;
  try {
    decoded = new TextDecoder('utf-8', { fatal: true }).decode(payload);
  } catch {
    throw new TerminalProtocolError(1008, 'Terminal JSON payload is not UTF-8');
  }
  let value: unknown;
  try {
    value = JSON.parse(decoded);
  } catch {
    throw new TerminalProtocolError(1008, 'Terminal JSON payload is invalid');
  }
  return asRecord(value);
}

function asRecord(value: unknown): Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new TerminalProtocolError(1008, 'Terminal JSON payload must be an object');
  }
  return value as Record<string, unknown>;
}

function assertExactKeys(value: Record<string, unknown>, expected: readonly string[]): void {
  const keys = Object.keys(value);
  if (
    keys.length !== expected.length ||
    expected.some((key) => !Object.prototype.hasOwnProperty.call(value, key))
  ) {
    throw new TerminalProtocolError(1008, 'Terminal JSON schema contains unknown fields');
  }
}

function toBytes(data: unknown): Uint8Array {
  if (data instanceof ArrayBuffer) {
    return new Uint8Array(data);
  }
  if (ArrayBuffer.isView(data)) {
    return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
  }
  throw new TerminalProtocolError(1008, 'Binary terminal frames are required');
}

function withChannel(channel: number, payload: Uint8Array): Uint8Array {
  const frame = new Uint8Array(payload.byteLength + 1);
  frame[0] = channel;
  frame.set(payload, 1);
  return frame;
}

File: CubeOps/internal/terminal/protocol.go
// Copyright (c) 2026 Tencent Inc.
// SPDX-License-Identifier: Apache-2.0

// Package terminal implements the browser-facing cube-terminal.v1 protocol
// and the stateless CubeOps WebSocket gateway.
package terminal

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"

	"github.com/gorilla/websocket"
)

const (
	Subprotocol = "cube-terminal.v1"
	GrantPrefix = "cube-grant."

	ChannelStdin  byte = 0x00
	ChannelStdout byte = 0x01
	ChannelStderr byte = 0x02
	ChannelStatus byte = 0x03
	ChannelResize byte = 0x04

	maxStatusBytes = 4 << 10
)

var terminalErrorCodes = map[string]struct{}{
	"TARGET_NOT_FOUND":   {},
	"TARGET_NOT_RUNNING": {},
	"LIMIT_EXCEEDED":     {},
	"PROTOCOL_ERROR":     {},
	"SHELL_NOT_FOUND":    {},
	"SLOW_PRODUCER":      {},
	"SLOW_CONSUMER":      {},
	"SESSION_LOST":       {},
	"INTERNAL":           {},
	"SERVER_DRAINING":    {},
	"SANDBOX_TRANSITION": {},
}

var terminalCloseReasons = map[string]struct{}{
	"USER_CLOSED":        {},
	"IDLE_TIMEOUT":       {},
	"MAX_LIFETIME":       {},
	"SANDBOX_TRANSITION": {},
	"RUNTIME_EXITED":     {},
	"SESSION_LOST":       {},
	"PROTOCOL_ERROR":     {},
	"SERVER_DRAINING":    {},
	"SLOW_PRODUCER":      {},
	"SLOW_CONSUMER":      {},
	"INTERNAL":           {},
}

type ProtocolError struct {
	CloseCode int
	Message   string
}

func (e *ProtocolError) Error() string { return e.Message }

type ClientFrameInfo struct {
	StdinBytes int64
}

type ServerFrameInfo struct {
	Type        string
	SessionID   string
	StdoutBytes int64
	ExitCode    *int32
	ErrorCode   string
	CloseReason string
}

type resizePayload struct {
	Cols uint32 `json:"cols"`
	Rows uint32 `json:"rows"`
}

type replayPayload struct {
	From      uint64 `json:"from"`
	Truncated bool   `json:"truncated"`
}

type statusPayload struct {
	Type      string         `json:"type"`
	SessionID string         `json:"sessionId,omitempty"`
	Replay    *replayPayload `json:"replay,omitempty"`
	ExitCode  *int32         `json:"exitCode,omitempty"`
	Code      string         `json:"code,omitempty"`
	Reason    string         `json:"reason,omitempty"`
}

func ValidateClientMessage(messageType int, message []byte, maxFrameBytes int) (ClientFrameInfo, error) {
	if messageType != websocket.BinaryMessage {
		return ClientFrameInfo{}, &ProtocolError{CloseCode: websocket.ClosePolicyViolation, Message: "binary terminal frames required"}
	}
	if maxFrameBytes <= 0 || len(message) == 0 {
		return ClientFrameInfo{}, &ProtocolError{CloseCode: websocket.ClosePolicyViolation, Message: "terminal frame is empty"}
	}
	if len(message)-1 > maxFrameBytes {
		return ClientFrameInfo{}, &ProtocolError{CloseCode: websocket.CloseMessageTooBig, Message: "terminal frame exceeds limit"}
	}

	payload := message[1:]
	switch message[0] {
	case ChannelStdin:
		return ClientFrameInfo{StdinBytes: int64(len(payload))}, nil
	case ChannelResize:
		if len(payload) == 0 || len(payload) > maxStatusBytes {
			return ClientFrameInfo{}, &ProtocolError{CloseCode: websocket.ClosePolicyViolation, Message: "terminal resize payload is invalid"}
		}
		var resize resizePayload
		if err := decodeStrictJSON(payload, &resize); err != nil {
			return ClientFrameInfo{}, &ProtocolError{CloseCode: websocket.ClosePolicyViolation, Message: "terminal resize payload is invalid"}
		}
		if resize.Cols == 0 || resize.Rows == 0 || resize.Cols > 1000 || resize.Rows > 1000 {
			return ClientFrameInfo{}, &ProtocolError{CloseCode: websocket.ClosePolicyViolation, Message: "terminal dimensions are out of range"}
		}
		return ClientFrameInfo{}, nil
	default:
		return ClientFrameInfo{}, &ProtocolError{
			CloseCode: websocket.ClosePolicyViolation,
			Message:   fmt.Sprintf("terminal channel 0x%02x is not valid from the client", message[0]),
		}
	}
}

func ValidateServerMessage(messageType int, message []byte, maxFrameBytes int) (ServerFrameInfo, error) {
	if messageType != websocket.BinaryMessage {
		return ServerFrameInfo{}, errors.New("terminal relay returned a non-binary frame")
	}
	if maxFrameBytes <= 0 || len(message) == 0 {
		return ServerFrameInfo{}, errors.New("terminal relay returned an empty frame")
	}
	payload := message[1:]
	switch message[0] {
	case ChannelStdout:
		if len(payload) > maxFrameBytes {
			return ServerFrameInfo{}, errors.New("terminal stdout frame exceeds limit")
		}
		return ServerFrameInfo{Type: "stdout", StdoutBytes: int64(len(payload))}, nil
	case ChannelStatus:
		if len(payload) == 0 || len(payload) > maxStatusBytes {
			return ServerFrameInfo{}, errors.New("terminal status frame is invalid")
		}
		var status statusPayload
		if err := decodeStrictJSON(payload, &status); err != nil {
			return ServerFrameInfo{}, fmt.Errorf("decode terminal status: %w", err)
		}
		return validateStatus(status)
	default:
		return ServerFrameInfo{}, fmt.Errorf("terminal relay returned invalid channel 0x%02x", message[0])
	}
}

func validateStatus(status statusPayload) (ServerFrameInfo, error) {
	info := ServerFrameInfo{Type: status.Type}
	switch status.Type {
	case "opened":
		if status.SessionID == "" || status.Replay == nil || status.ExitCode != nil || status.Code != "" || status.Reason != "" {
			return ServerFrameInfo{}, errors.New("terminal opened status schema is invalid")
		}
		info.SessionID = status.SessionID
	case "exit":
		if status.ExitCode == nil || status.SessionID != "" || status.Replay != nil || status.Code != "" || status.Reason != "" {
			return ServerFrameInfo{}, errors.New("terminal exit status schema is invalid")
		}
		info.ExitCode = status.ExitCode
	case "error":
		if status.Code == "" || status.SessionID != "" || status.Replay != nil || status.ExitCode != nil || status.Reason != "" {
			return ServerFrameInfo{}, errors.New("terminal error status schema is invalid")
		}
		if _, ok := terminalErrorCodes[status.Code]; !ok {
			return ServerFrameInfo{}, errors.New("terminal error status code is unknown")
		}
		info.ErrorCode = status.Code
	case "close":
		if status.Reason == "" || status.SessionID != "" || status.Replay != nil || status.ExitCode != nil || status.Code != "" {
			return ServerFrameInfo{}, errors.New("terminal close status schema is invalid")
		}
		if _, ok := terminalCloseReasons[status.Reason]; !ok {
			return ServerFrameInfo{}, errors.New("terminal close status reason is unknown")
		}
		info.CloseReason = status.Reason
	default:
		return ServerFrameInfo{}, errors.New("terminal status type is unknown")
	}
	return info, nil
}

func EncodeErrorStatus(code string) ([]byte, error) {
	if code == "" {
		return nil, errors.New("terminal error code is empty")
	}
	return encodeStatus(statusPayload{Type: "error", Code: code})
}

func EncodeCloseStatus(reason string) ([]byte, error) {
	if reason == "" {
		return nil, errors.New("terminal close reason is empty")
	}
	return encodeStatus(statusPayload{Type: "close", Reason: reason})
}

func encodeStatus(status statusPayload) ([]byte, error) {
	payload, err := json.Marshal(status)
	if err != nil {
		return nil, err
	}
	if len(payload) > maxStatusBytes {
		return nil, errors.New("terminal status exceeds limit")
	}
	message := make([]byte, 1+len(payload))
	message[0] = ChannelStatus
	copy(message[1:], payload)
	return message, nil
}

func decodeStrictJSON(payload []byte, destination any) error {
	decoder := json.NewDecoder(bytes.NewReader(payload))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(destination); err != nil {
		return err
	}
	if err := decoder.Decode(&struct{}{}); !errors.Is(err, io.EOF) {
		if err == nil {
			return errors.New("multiple JSON values are not allowed")
		}
		return err
	}
	return nil
}
